### PR TITLE
Discord action roles

### PR DESCRIPTION
**Pull request by DashDeipayan**
This PR contains the code for the features of [create-group-roles](https://github.com/Real-Dev-Squad/website-dashboard/issues/340) and [add-group-role](https://github.com/Real-Dev-Squad/website-dashboard/issues/341) and [isDiscordVerified](https://github.com/Real-Dev-Squad/website-backend/issues/1062) for `website dashboard` integration. 

# N.B. This PR is NOT TO BE MERGED before [this issue #1061](https://github.com/Real-Dev-Squad/website-backend/issues/1061) is Resolved!!
- [x] https://github.com/Real-Dev-Squad/website-backend/issues/1061

### CODE REVIEW

File: controllers/discordactions.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+const createGroupRole = async (req, res) => {
+  try {
+    const rolename = `group-${req.body.rolename}`;
+
+    const { wasSuccess } = await discordRolesModel.isGroupRoleExists(rolename);
+
+    if (!wasSuccess) {
+      return res.status(400).json({
+        message: "Role already exists!",
+      });
```
Comment (RitikJaiswal75) [inline controllers/discordactions.js:25]: what if it fails for some other reason, the message being sent is for HTTP status 409
are we sending 400 irrespective of the reason?

File: controllers/discordactions.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      mentionable: true,
+    };
+    const groupRoleData = {
+      rolename,
+      createdBy: req.userData.id,
+      date: admin.firestore.Timestamp.fromDate(new Date()),
+    };
+    const authToken = jwt.sign({}, config.get("rdsServerlessBot.rdsServerLessPrivateKey"), {
+      algorithm: "RS256",
+      expiresIn: config.get("userToken.ttl"),
```
Comment (RitikJaiswal75) [inline controllers/discordactions.js:38]: let's keep this for a smaller duration, please. userToken exists for a longer duration.

Our cloudflare bot will only be active for 10 ms so I think the max age of token can be 1 minute

File: controllers/discordactions.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    logger.error(`Error while getting roles: ${err}`);
+    return res.boom.badImplementation(INTERNAL_SERVER_ERROR);
+  }
+};
+
+/**
+ * Gets all group-roles
+ *
+ * @param res {Object} - Express response object
+ */
```
Comment (RitikJaiswal75) [inline controllers/discordactions.js:82]: change this to be appropriate please

File: controllers/discordactions.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    const { roleData, wasSuccess } = await discordRolesModel.addGroupRoleToMember(memberGroupRole);
+
+    if (!wasSuccess) {
+      return res.status(400).json({
+        message: "Role already exists!",
+        data: {
+          ...roleData,
+        },
+      });
+    }
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply controllers/discordactions.js:100]: if it is adding a role to a member, wont they be choosing from the pre-created roles? why do we get a role already exists message here?

File: controllers/discordactions.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    const dataForDiscord = {
+      ...req.body,
+    };
+    const authToken = jwt.sign({}, config.get("rdsServerlessBot.rdsServerLessPrivateKey"), {
+      algorithm: "RS256",
+      expiresIn: config.get("userToken.ttl"),
```
Comment (RitikJaiswal75) [inline controllers/discordactions.js:106]: reduce the max age

File: controllers/discordactions.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    await fetch(`${DISCORD_BASE_URL}/roles/add`, {
+      method: "PUT",
+      body: JSON.stringify(dataForDiscord),
+      headers: { "Content-Type": "application/json", Authorization: `Bearer ${authToken}` },
+    }).then((response) => response.json());
+
+    return res.json({
+      message: "Role added successfully!",
+    });
```
Comment (RitikJaiswal75) [inline controllers/discordactions.js:117]: status code please

File: controllers/discordactions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply controllers/discordactions.js:100]: expecting correct behaviour from user is not ideal. We must create validations for edge cases

File: controllers/discordactions.js (hunk 6 above)
Comment (DashDeipayan) [reply controllers/discordactions.js:117]: adding

File: controllers/discordactions.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      body: JSON.stringify(dataForDiscord),
+      headers: { "Content-Type": "application/json", Authorization: `Bearer ${authToken}` },
+    }).then((response) => response.json());
+
+    groupRoleData.roleid = responseForCreatedRole.id;
+
+    await discordRolesModel.createNewRole(groupRoleData);
+    return res.json({
+      message: "Role created successfully!",
+    });
```
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply controllers/discordactions.js:52]: adding

File: config/test.js [hunk 8]
```diff
@@ -102,4 +105,45 @@ module.exports = {
       "Ba+XMCP64pXPC3r1llhKRwIl+6UFn+QlpbxtgQjhbULnSbc7fw==\n" +
       "-----END RSA PRIVATE KEY-----",
   },
+
+  rdsServerlessBot: {
+    rdsServerLessPublicKey:
+      "-----BEGIN PUBLIC KEY-----\n" +
```
Comment (bhtibrewal) [inline config/test.js:111]: remove this key and add a dummy key here

File: controllers/discordactions.js (hunk 3 above)
Comment (DashDeipayan) [reply controllers/discordactions.js]: fixed

**Review by bhtibrewal — DISMISSED** [review]
LGTM

**Review by pallabez — DISMISSED** [review]
Thanks for the PR. LGTM

File: config/default.js [hunk 9]
```diff
@@ -78,6 +78,11 @@ module.exports = {
     CLOUDFLARE_X_AUTH_EMAIL: "Cloudflare_User_Email",
   },
 
+  rdsServerlessBot: {
+    rdsServerLessPrivateKey: "RDS_SERVERLESS_PRIVATE_KEY",
+    ttl: "RDS_SERVERLESS",
```
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply config/default.js:83]: Thats a `default.js` file so I should update it with some numeric value. But thanks that you caught it.

File: controllers/discordactions.js [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      mentionable: true,
+    };
+    const groupRoleData = {
+      rolename,
+      createdBy: req.userData.id,
+      date: admin.firestore.Timestamp.fromDate(new Date()),
+    };
+    const authToken = jwt.sign({}, config.get("rdsServerlessBot.rdsServerLessPrivateKey"), {
+      algorithm: "RS256",
+      expiresIn: config.get("userToken.ttl"),
```
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply controllers/discordactions.js:38]: Oops, missed it. Thanks.

**Review by RitikJaiswal75 — APPROVED** [review]
LGTM

**Review by bhtibrewal — APPROVED** [review]
Okay to merge

**Pull request merged**